patches/tModLoader/Terraria/ModLoader/AtmosphericLoader.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System.Collections.Generic;
+using Terraria.Graphics.Capture;
+
+namespace Terraria.ModLoader
+{
+	//todo: further documentation
+	/// <summary>
+	/// This serves as the central class from which AVFX functions are supported and carried out.
+	/// </summary>
+	public abstract class AVFXLoader<T>
+	{
+		public int vanillaCount { get; internal set; }
+		public int totalCount { get; internal set; }
+
+		internal List<T> list = new List<T>();
+
+		protected AVFXLoader(int vanillaCount) {
+			this.vanillaCount = vanillaCount;
+			totalCount = vanillaCount;
+		}
+
+		public int Reserve() {
+			int reserve = totalCount;
+			totalCount++;
+			return reserve;
+		}
+
+		public T Get(int id) {
+			if (id < vanillaCount || id >= totalCount) {
+				return default(T);
+			}
+			return list[id - vanillaCount];
+		}
+
+		internal void Unload() {
+			totalCount = vanillaCount;
+			list.Clear();
+		}
+
+		internal virtual void ResizeArrays() { }
+
+		public virtual void ChooseStyle(ref int style) { }
+	}
+
+	public static class AVFXHelper {
+		public static WaterFallStyles Waterfalls = new WaterFallStyles();
+		public static WaterStyles Waters = new WaterStyles();
+		public static UgBgStyles UgBg = new UgBgStyles();
+		public static SurfaceBgStyles SurfaceBg = new SurfaceBgStyles();
+	}
+
+	public static class AVFXLoader
+	{
+		internal static List<ModAVFX> AVFXs = new List<ModAVFX>();
+
+		public class ModAtmosphere
+		{
+			public bool anyActive;
+			public AVFXPriority priority;
+			public int waterStyle;
+			public int ugBG;
+			public int surfaceBG;
+			public CaptureBiome.TileColorStyle tileColorStyle;
+			public int music;
+
+			public ModAtmosphere() {
+				anyActive = false;
+				priority = AVFXPriority.None;
+				waterStyle = ugBG = surfaceBG = 0;
+				music = -1;
+				tileColorStyle = CaptureBiome.TileColorStyle.Normal;
+			}
+		}
+
+		internal struct AtmosWeight{
+			internal int weight;
+			internal ModAVFX type;
+
+			internal AtmosWeight(int weight, ModAVFX type) {
+				this.weight = weight;
+				this.type = type;
+			}
+		}
+
+		public static void UpdateModAtmosphere(Player player) {
+			var result = new ModAtmosphere();
+			List<AtmosWeight> shortList = new List<AtmosWeight>();
+
+			foreach (var atmos in AVFXs) {
+				int corrWeight = atmos.GetCorrWeight(player);
+				if (corrWeight == 0)
+					continue;
+
+				shortList.Add(
+					new AtmosWeight(corrWeight, atmos)
+				);
+			}
+
+			if (!(shortList.Count == 0)) {
+				result.anyActive = true;
+				int weight = 0;
+
+				shortList.Sort((s1, s2) => s1.weight.CompareTo(s2.weight));
+
+				foreach (var atmos in shortList) {
+					if (atmos.weight > weight) {
+						weight = atmos.weight;
+
+						result.priority = atmos.type.Priority;
```

Comment:
Agrair found bug:
biome: priority 1: music X1, background X2
event: priority 4: background X3.
Result: priority 4: background X3, music X1
Which means that either would have to wrap vanilla and deviate from how ModSystem did the ChooseMusic originally OR assume all derived classes have music. Notably, this only applies to music selection against vanilla today.
Given that the priority result is just cached, no reason we couldn't set it to match music at this point in time?

Reply:
As a hotfix and possibly a permanent change, I did just [that](https://github.com/tModLoader/tModLoader/pull/1205/files#diff-9a1f12f29160d4c2b7c6a77b3ff9a11217e584089609fb35582ec0778bccfae3R140). However, we should leave this thread open as it is liable to change.